src/geometry/manifold/derivation_bundle.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import geometry.manifold.algebra.smooth_functions
+import ring_theory.derivation
+
+/-!
+
+# Derivation bundle
+
+In this file we define the derivations at a point of a manifold on the algebra of smooth fuctions.
+Moreover we define the total bundle of derivations (although at the moment it has not been given a
+topology). Finally we define the differential of a function in terms of derivations.
+
+The content of this file is not meant to be regarded as an alternative definition to the current
+tangent bundle but rather as a purely algebraic theory that provides a purely algebraic definition
+of the Lie algebra for a Lie group.
+
+-/
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+{M : Type*} [topological_space M] [charted_space H M]
+
+open_locale manifold
+
+namespace point_derivation
+
+instance smooth_functions_algebra : algebra 𝕜 C^∞⟮I, M; 𝕜⟯ := by apply_instance
+instance smooth_functions_tower : is_scalar_tower 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ := by apply_instance
+
+/-- Evaluation at a point is a ring homomorphism. -/
+def smooth_function.eval' (x : M) : C^∞⟮I, M; 𝕜⟯ →+* 𝕜 :=
+{ to_fun    := λ f, f x,
+  map_one'  := rfl,
+  map_mul'  := λ f g, rfl,
+  map_zero' := rfl,
+  map_add'  := λ f g, rfl }
```

Comment:
```suggestion
/-- Evaluation at a point is a ring homomorphism. -/
def smooth_function.eval_ring_hom (x : M) : C^∞⟮I, M; 𝕜⟯ →+* 𝕜 :=
{ to_fun    := λ f, f x,
  map_one'  := rfl,
  map_mul'  := λ f g, rfl,
  map_zero' := rfl,
  map_add'  := λ f g, rfl }
```
which is more consistent with `continuous_map.coe_fn_ring_hom`.

Ideally, you could instead define
```suggestion
/-- Coercion is a ring homomorphism. Similar to `continuous_map.coe_fn_ring_hom`. -/
def smooth_function.coe_fn_ring_hom : C^∞⟮I, M; 𝕜⟯ →+* (M → 𝕜) :=
{ to_fun    := coe_fn,
  map_one'  := coe_one,
  map_mul'  := coe_mul,
  map_zero' := coe_zero,
  map_add'  := coe_add }

/-- Evaluation at a point is a ring homomorphism. -/
def smooth_function.eval_ring_hom (x : M) : C^∞⟮I, M; 𝕜⟯ →+* 𝕜 :=
pi.eval_ring_hom.comp smooth_function.coe_fn_ring_hom
```
That probably doesn't work quite as written, but it should be straightforward to resolve.

Reply:
Yes I was planning to do this but I was waiting #7749 to be merged not to bother creating dependencies...

Reply:
Actually this does not work because `pi.eval_ring_hom` gives the error
```
invalid field notation, type is not of the form (C ...) where C is a constant
  pi.eval_ring_hom
has type
  Π (f : ?m_1 → Type ?) [_inst_1 : Π (i : ?m_1), non_assoc_semiring (f i)] (i : ?m_1),  (Π (i : ?m_1), f i) →+* f i
```
and I am not really sure how to solve this.

Note that it is `smooth_map` now and not `smooth_function`. But if you manage to send me a working snippet for the analogue in the case of `continuous_map.coe_fn_ring_hom` I will imitate that.

Reply:
You probably need `(pi.eval_ring_hom _ : (M → 𝕜) →+* 𝕜).comp` to fix that error

Reply:
This is what I mean:
```suggestion
/-- Evaluation at a point is a ring homomorphism. -/
def smooth_map.eval_ring_hom (x : M) : C^∞⟮I, M; 𝕜⟯ →+* 𝕜 :=
(pi.eval_ring_hom _ x : (M → 𝕜) →+* 𝕜).comp smooth_map.coe_fn_ring_hom
```

Reply:
I am not sure I want to put it outside though because this evaluation makes sense only in the context of point derivations I believe.

Reply:
You should put it outside - it makes just as much sense as `pi.eval_ring_hom`.

Also, I'm confused by your use of `smooth_function` in names here - doesn't mathlib only have `smooth_map`?